### PR TITLE
[TEC-3477] Localized order details for Spree Admin, payments mapping,  OrderUpdater service, refunds

### PR DESCRIPTION
**Pull request by texpert**
### What problem is the code solving?

Using the localized data from flow.io for displaying order details for the Spree Admin site and emails

### How does this change address the problem?

Adjusted the display_total, flow_order, flow_line_item_price, flow_total (renamed to flow_io_total_amount).

Removed obsolete OrderSync.total_price method and Spree::LineItem.single_money method - the latter is perfectly working from the main app.

### Share the knowledge
Changed Spree::Order decorator to be included as FlowIoOrderDecorator:
 - we need the `display_total` method to work with flow_data, but using `class_eval` the decorator defined in `flowcommerce_spree` is overridden by the main app's Spree::Order decorator. The try to prepend the decorator module lead to Spree::Order's state_machine breakage. That's why the final solution it is to include the module and add a call to `super` in the main app's `display_total` method

### CODE REVIEW

File: app/controllers/flowcommerce_spree/orders_controller.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+module FlowcommerceSpree
+  class OrdersController < ApplicationController
+    wrap_parameters false
+
+    skip_before_action :setup_tracking, only: :order_completed
+
```
Comment (davidmejuri) [inline app/controllers/flowcommerce_spree/orders_controller.rb:8]: Hi @texpert Why are we adding this endpoint on this Pull Request? It looks like the endpoint for validating the puchase and not dedicated to the spree admin

File: app/models/spree/flow_io_credit_card_decorator.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    end
+
+    def push_authorization(auth_hash)
+      self.flow_data ||= {}
+      flow_data['authorizations'] ||= []
+      card_authorizations = flow_data['authorizations']
+      card_authorizations.delete_if { |ca| ca['id'] == auth_hash['id'] }
+      card_authorizations << auth_hash
+    end
+
```
Comment (davidmejuri) [inline app/models/spree/flow_io_credit_card_decorator.rb:18]: @texpert Are we using Spree::CreditCard modeling for handling payments? If it generates problem we. could create a new payment method dedicated to flow

File: app/models/spree/flow_io_order_decorator.rb [hunk 3]
```diff
@@ -1,15 +1,16 @@
...
+  module FlowIoOrderDecorator
+    def self.included(base)
+      base.serialize :meta, ActiveRecord::Coders::JSON.new(symbolize_keys: true)
 
-    store_accessor :meta, :flow_data
+      base.store_accessor :meta, :flow_data
 
-    before_save :sync_to_flow_io
-    after_touch :sync_to_flow_io
+      base.before_save :sync_to_flow_io
```
Comment (davidmejuri) [inline app/models/spree/flow_io_order_decorator.rb:11]: @texpert Aren't we synchronizing the order twice when `.save` method is called?

File: app/models/spree/flow_io_order_decorator.rb [hunk 4]
```diff
@@ -172,8 +169,27 @@ def flow_order_authorized?
...
 
-    def flow_order_captured?
-      flow_data['capture'] ? true : false
+    def flow_io_captures
+      flow_data&.[]('captures')
+    end
+
+    def flow_io_captures_sum
+      captures_sum = 0
+      # flow_data&.[]('captures')&.select { |c| c['status'] == 'succeeded' }&.map { |c| c['amount'] }&.sum.to_d
```
Comment (davidmejuri) [inline app/models/spree/flow_io_order_decorator.rb:178]: Please delete commented code 🙏

File: app/services/flowcommerce_spree/order_updater.rb [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        payment.pend
+
+        # For now this additional update is overwriting the generated identifier with flow.io payment identifier.
+        # TODO: Check and possibly refactor in Spree 3.0, where the `before_create :set_unique_identifier`
+        # has been removed.
+        payment.update_column(:identifier, p['id'])
+      end
+
+      return if @order.payments.sum(:amount) < @order.amount || @order.state == 'complete'
+
```
Comment (davidmejuri) [inline app/services/flowcommerce_spree/order_updater.rb:66]: Can't we just use `order.next!`?

File: app/controllers/flowcommerce_spree/orders_controller.rb (hunk 1 above)
Comment (texpert) [reply app/controllers/flowcommerce_spree/orders_controller.rb:8]: Hi, @davidmejuri !

This PR will be renamed - it is the Phase_3 release, in fact.

File: app/models/spree/flow_io_credit_card_decorator.rb (hunk 2 above)
Comment (texpert) [reply app/models/spree/flow_io_credit_card_decorator.rb:18]: We're not handling payments, we're just mapping the received info from flow.io into Spree, so everything will continue to be flawlessly processed - displaying of payments and captures in Spree::Admin, refunding operations, anything.

File: app/models/spree/flow_io_order_decorator.rb (hunk 3 above)
Comment (texpert) [reply app/models/spree/flow_io_order_decorator.rb:11]: Yeah, this should be definitely refactored - I have created an issue for this - https://github.com/mejuri-inc/flowcommerce_spree/issues/23

File: app/models/spree/flow_io_order_decorator.rb (hunk 4 above)
Comment (texpert) [reply app/models/spree/flow_io_order_decorator.rb:178]: Yeah, it is deleted in the next PR targeted onto this branch - https://github.com/mejuri-inc/flowcommerce_spree/pull/25/files#diff-e683f17a9e0ac64e1001e420f4937ec3760eb6717e1707acb41d51c9fa2d446eL178

File: app/services/flowcommerce_spree/order_updater.rb (hunk 5 above)
Comment (texpert) [reply app/services/flowcommerce_spree/order_updater.rb:66]: Nope. Just have double-checked and the order's state remains `confirm` after 2 `next` calls.

**Review by davidmejuri — APPROVED** [review]

**Review by sebastiandl — APPROVED** [review]

**Pull request merged**